lib/client.js

```diff
@@ -254,16 +259,44 @@ Object.assign(Client.prototype, {
     // choose the smallest, positive timeout setting and set on poolConfig
     poolConfig.acquireTimeoutMillis = Math.min(...timeouts);
 
+    const updatePoolConnectionSettingsFromProvider = () => {
+      if (!this.connectionConfigProvider) {
+        return; // static configuration, nothing to update
+      }
+      if (
+        this.connectionSettings.expired === undefined ||
+        !this.connectionSettings.expired()
+      ) {
+        return; // reuse existing connection
+      }
+      const providerResult = this.connectionConfigProvider();
+      if (!providerResult.then) {
```

Comment:
why can't you just always `await` it?

Reply:
This allows the provider to be either synchronous or async.

Reply:
it's perfectly fine to await synchronous functions.

Reply:
According to the [doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await):
If the value is not a Promise, it converts the value to a resolved Promise, and waits for it.

It sounds like this might split execution flow, causing a slight performance impact, especially on older Node versions (I think this has been optimized away in Node 12). For this reason I was trying not to add any async stuff along the core execution path unless it's required for supporting the new functionality.

Just bringing this consideration to attention - if you still feel it's better to use await, no problem I will change it.

Reply:
@oranoran Please correct me if I'm wrong, but this does not seem to be a hot path and is likely to be executed fairly rarely, hence microoptimization unlikely to be useful, and it does simplify code quite a bit in this case.